### PR TITLE
add outputPath Config

### PR DESCRIPTION
**Pull request by ringcrl**
add outputPath config for leetcode #119

Change Request,Need Help...Find a way to get a Problem difficulty...

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -114,13 +114,17 @@
...
 
 ## Settings
-| Setting Name | Description | Default Value |
-|---|---|---|
-| `leetcode.hideSolved` | Specify to hide the solved problems or not | `false` |
-| `leetcode.showLocked` | Specify to show the locked problems or not. Only Premium users could open the locked problems | `false` |
-| `leetcode.defaultLanguage` | Specify the default language used to solve the problem. Supported languages are: `bash`, `c`, `cpp`, `csharp`, `golang`, `java`, `javascript`, `kotlin`, `mysql`, `python`,`python3`,`ruby`,`scala`,`swift` | `N/A` |
-| `leetcode.useWsl` | Specify whether to use WSL or not | `false` |
-| `leetcode.endpoint` | Specify the active endpoint. Supported endpoints are: `leetcode`, `leetcode-cn` | `leetcode` |
+| Setting Name               | Description                                                                                                                                                                                                 | Default Value |
```
Comment (jdneo) [inline README.md]: Hmm, please revert the changes to the markdown document. Let me do this.

**Comment by jdneo** [comment]
@ringcrl 

I think there is still some misunderstanding here. The aim of add the setting `leetcode.outputPath` is to make it possible for the users to categorize the files as their wish. So instead of make it as a enum setting, we should make it as a plain string.

Let me explain more about it:

The setting of `leetcode.outputPath` has no default value, which means it is empty by default. And, by default, when the user click `show problem`, the file goes to the base path of the workspace folder.

If the user set some values to it, for example: 'happy' then the file goes to the folder called `happy`

Meanwhile, besides let the user customize the output folder, we will also provide some **special** values, which are, as you know, `${tag}`, `${language}`, `${difficulty}`. This means the user needs to explicitly brace the value with `${}`, then the extension will get to know that: Oh, this is not a general path. I need to traslate it to another value...

Do you get my point?

File: src/commands/show.ts [hunk 2]
```diff
@@ -49,7 +49,25 @@ async function showProblemInternal(id: string): Promise<void> {
...
 
-        const outDir: string = await selectWorkspaceFolder();
+        let outDir: string = await selectWorkspaceFolder();
+        const outputPath: string = leetCodeConfig.get<string>("outputPath") || "root";
+        switch (outputPath) {
+            case "root": {
+                break;
+            }
+            case "tag": {
+                const { tags } = await leetCodeExecutor.getCompaniesAndTags();
```
Comment (jdneo) [inline src/commands/show.ts]: `Tag` and `Difficulty` is stored in the `LeetCodeNode`. 

See: https://github.com/jdneo/vscode-leetcode/blob/master/src/explorer/LeetCodeNode.ts#L28-L34
https://github.com/jdneo/vscode-leetcode/blob/master/src/commands/show.ts#L15-L19

**Comment by ringcrl** [comment]
Thank you very much, I have made changes.

File: src/commands/show.ts [hunk 3]
```diff
@@ -49,7 +50,34 @@ async function showProblemInternal(id: string): Promise<void> {
             return;
         }
 
-        const outDir: string = await selectWorkspaceFolder();
+        let outDir: string = await selectWorkspaceFolder();
+        const outputPath: string = leetCodeConfig.get<string>("outputPath") || "";
+        const problem: IProblem | undefined = listProblems.find((item: IProblem) => item.id === id);
+        switch (outputPath) {
```
Comment (jdneo) [inline src/commands/show.ts]: Steps to treat with the `outputPath`:

## Step 1: test if it's braced by `${}`:
> You can do this using regex: `\$\{(.*?)\}`, the regex can help you extract the value inside the braces

If yes, translate for `tag`, `language`, `difficulty`

### Why?
1. If the user provides a value which is not supported yet, for example: `${company}`, we will have a chance to tell the user this category is not supported yet
2. Use `switch(value.toLocaleLowerCase())` to make sure if the user provides `${Difficulty}`, the feature will still work.

## Step 2: join the path
If it's not braced by `${}`, then we can directly join the path without translation.

There is no need to treat with empty value because:

1. `leetCodeConfig.get<string>("outputPath") || ""` make sure the value will not be undefined
2. `path.join("a", "")` will ignore the empty string

**Comment by ringcrl** [comment]
Thank you very much for these helpful suggestions.

File: src/commands/show.ts [hunk 4]
```diff
@@ -34,10 +36,10 @@ export async function searchProblem(): Promise<void> {
     if (!choice) {
         return;
     }
-    await showProblemInternal(choice.value);
+    await showProblemInternal(problems.find((problem: IProblem) => problem.id === choice.value) as IProblem);
```
Comment (jdneo) [inline src/commands/show.ts]: Have you tried to change the line 118 from 
`value: problem.id,` to `value: problem`?

File: src/commands/show.ts [hunk 5]
```diff
@@ -16,16 +17,17 @@ export async function showProblem(node?: LeetCodeNode): Promise<void> {
     if (!node) {
         return;
     }
-    await showProblemInternal(node.id);
+    await showProblemInternal(node);
```
Comment (jdneo) [inline src/commands/show.ts]: Though type casting from `LeetCodeNode` to `IProblem` is fine. Let's add getter in `LeetCodeNode` to get the data, which is typed as `IProblem`.

> It looks like I need to do some refactors here later after this PR, The definition of `LeetCodeNode` and `IProblem` is overlapped.

File: src/commands/show.ts [hunk 6]
```diff
@@ -49,9 +51,36 @@ async function showProblemInternal(id: string): Promise<void> {
...
         }
 
-        const outDir: string = await selectWorkspaceFolder();
+        let outDir: string = await selectWorkspaceFolder();
+        const outputPathCfg: string = leetCodeConfig.get<string>("outputPath") || "";
+        const outputPath: RegExpMatchArray | null = outputPathCfg.match(/\$\{(.*?)\}/);
+        if (outputPath) {
+            switch (outputPath[1].toLowerCase()) {
+                case "tag": {
+                    const closestTag: string = node.tags.reduce((prev: string, curr: string) => {
```
Comment (jdneo) [inline src/commands/show.ts]: Hmm, here is the problem, looks like you want to find out the tag with the longest length if the problem has multiple tags. Any reason for this logic?

File: src/commands/show.ts [hunk 7]
```diff
@@ -49,9 +51,36 @@ async function showProblemInternal(id: string): Promise<void> {
...
+                            prev;
+                    }, "");
+                    outDir = path.join(outDir, closestTag);
+                    break;
+                }
+                case "language": {
+                    outDir = path.join(outDir, language);
+                    break;
+                }
+                case "difficulty": {
```
Comment (jdneo) [inline src/commands/show.ts]: No need to add `{}` for each `case` and `default`

File: src/commands/show.ts (hunk 6 above)
Comment (ringcrl) [reply src/commands/show.ts]: I found that if I took the first one, most of them are [array]. This is not what I want. By observing the all of tags, I found that the longest tag is generally the best one...I originally wanted to add a priority to the tag, which is complicated.

File: src/commands/show.ts (hunk 4 above)
Comment (ringcrl) [reply src/commands/show.ts]: emm...Because of type error, so I use IProblem on both sides.

File: src/commands/show.ts (hunk 6 above)
Comment (jdneo) [reply src/commands/show.ts]: Then, this is not convincing enough. Please considering:
- We have no idea what does user want
- There is no data to support the conclusion: `the longest tag is generally the best one`, it's just a subjective judgment

Meanwhile, by doing this, we provided an implicit logic for the user, this is not a recommended user experience.

I have two suggestions:
1. show quick picks to the user if multiple tags are available, let the user choose one.
2. Remove the support for `${tag}` for now.

File: src/commands/show.ts (hunk 4 above)
Comment (jdneo) [reply src/commands/show.ts]: I mean, you are using `problems.find((problem: IProblem) => problem.id === choice.value)` to get the problem, that is because the `choice.value` only stores the `id`. You can make it to store the problem itself.

File: src/commands/show.ts (hunk 6 above)
Comment (ringcrl) [reply src/commands/show.ts]: Thanks for your suggestion，This solution is very reasonable

**Comment by ringcrl** [comment]
Add tag selector

**Comment by jdneo** [comment]
@ringcrl Looks much better now.

I just did some changes and merged the PR.

Thank you for your contribution.
